src/platform/datapath_linux.c

```diff
@@ -2114,22 +2084,64 @@ CxPlatSocketSend(
         }
     }
 
+    //
+    // Check to see if we need to pend.
+    //
+    if (!IsPendedSend) {
+        CxPlatLockAcquire(&SocketContext->PendingSendContextLock);
+        if (!CxPlatListIsEmpty(&SocketContext->PendingSendContextHead)) {
+            CxPlatSocketContextPendSend(
+                SocketContext,
+                SendData,
+                LocalAddress,
+                RemoteAddress);
+            SendPending = TRUE;
+        }
+        CxPlatLockRelease(&SocketContext->PendingSendContextLock);
+        if (SendPending) {
+            Status = QUIC_STATUS_PENDING;
+            goto Exit;
+        }
+    }
+
     SentByteCount = sendmsg(SocketContext->SocketFd, &Mhdr, 0);
 
     if (SentByteCount < 0) {
         if (errno == EAGAIN || errno == EWOULDBLOCK) {
-            Status =
+            if (!IsPendedSend) {
+                CxPlatLockAcquire(&SocketContext->PendingSendContextLock);
                 CxPlatSocketContextPendSend(
                     SocketContext,
                     SendData,
-                    ProcContext,
                     LocalAddress,
                     RemoteAddress);
-            if (QUIC_FAILED(Status)) {
-                goto Exit;
+                CxPlatLockRelease(&SocketContext->PendingSendContextLock);
             }
+            struct epoll_event SockFdEpEvt = {
+                .events = EPOLLIN | EPOLLOUT | EPOLLET,
+                .data = {
+                    .ptr = &SocketContext->EventContexts[QUIC_SOCK_EVENT_SOCKET]
+                }
+            };
 
+            int Ret =
+                epoll_ctl(
+                    ProcContext->EpollFd,
+                    EPOLL_CTL_MOD,
+                    SocketContext->SocketFd,
+                    &SockFdEpEvt);
+            if (Ret != 0) {
+                QuicTraceEvent(
+                    DatapathErrorStatus,
+                    "[data][%p] ERROR, %u, %s.",
+                    SocketContext->Binding,
+                    errno,
+                    "epoll_ctl failed");
+                Status = errno;
+                goto Exit;
+            }
             SendPending = TRUE;
```

Comment:
Need to set this above, even in the case of failure, so we don't free the send context on failure here, after we put it in the list already.

Reply:
Fixed.

Reply:
Ok